src/util/popup.ts

```diff
@@ -19,31 +25,47 @@ export class PopupService<T> {
 
   constructor(
       private _type: any, private _injector: Injector, private _viewContainerRef: ViewContainerRef,
-      private _renderer: Renderer2, private _componentFactoryResolver: ComponentFactoryResolver,
-      private _applicationRef: ApplicationRef) {}
+      private _renderer: Renderer2, private _ngZone: NgZone,
+      private _componentFactoryResolver: ComponentFactoryResolver, private _applicationRef: ApplicationRef) {}
 
-  open(content?: string | TemplateRef<any>, context?: any): ComponentRef<T> {
+  open(content?: string | TemplateRef<any>, context?: any, animation = false):
+      {windowRef: ComponentRef<T>, transition$: Observable<void>} {
     if (!this._windowRef) {
       this._contentRef = this._getContentRef(content, context);
       this._windowRef = this._viewContainerRef.createComponent(
           this._componentFactoryResolver.resolveComponentFactory<T>(this._type), this._viewContainerRef.length,
           this._injector, this._contentRef.nodes);
     }
 
-    return this._windowRef;
-  }
+    const {nativeElement} = this._windowRef.location;
+    const onStable$ = this._ngZone.onStable.asObservable().pipe(take(1));
+    const transition$ = onStable$.pipe(
+        mergeMap(
+            () => ngbRunTransition(
+                nativeElement, ({classList}) => classList.add('show'), {animation, runningTransition: 'continue'})));
 
-  close() {
-    if (this._windowRef) {
-      this._viewContainerRef.remove(this._viewContainerRef.indexOf(this._windowRef.hostView));
-      this._windowRef = null;
+    return {windowRef: this._windowRef, transition$};
+  }
 
-      if (this._contentRef?.viewRef) {
-        this._applicationRef.detachView(this._contentRef.viewRef);
-        this._contentRef.viewRef.destroy();
-        this._contentRef = null;
-      }
+  close(animation = false): Observable<any> {
+    if (!this._windowRef) {
+      return of(undefined);
     }
+
+    return ngbRunTransition(
+               this._windowRef.location.nativeElement, ({classList}) => classList.remove('show'),
+               {animation, runningTransition: 'continue'})
+        .pipe(tap(() => {
+          if (this._windowRef) {
+            this._viewContainerRef.remove(this._viewContainerRef.indexOf(this._windowRef.hostView));
```

Comment:
Can't we take the opportunity here to upgrade to `this._windowRef.hostView.destroy()` instead of `this._viewContainerRef.remove(this._viewContainerRef.indexOf(this._windowRef.hostView))` ?

Reply:
We could, sure. I wonder why it was not done before... it is already done for `this._contentRef.viewRef.destroy();` 5 lines below